dev/devicelab/lib/tasks/perf_tests.dart

```diff
@@ -320,13 +337,148 @@ class PerfTest {
         'average_vsync_transitions_missed',
         '90th_percentile_vsync_transitions_missed',
         '99th_percentile_vsync_transitions_missed',
-        if (needsMeasureCpuGPu) 'cpu_percentage',
-        if (needsMeasureCpuGPu) 'gpu_percentage',
+        if (needsMeasureCpuGpu) 'cpu_percentage',
+        if (needsMeasureCpuGpu) 'gpu_percentage',
       ]);
     });
   }
 }
 
+class PerfTestWithSkSL extends PerfTest {
+  PerfTestWithSkSL(
+    String testDirectory,
+    String testTarget,
+    String timelineFileName, {
+    bool needsMeasureCpuGpu = false,
+    String testDriver,
+  }) : super(
+    testDirectory,
+    testTarget,
+    timelineFileName,
+    needsMeasureCpuGpu: needsMeasureCpuGpu,
+    testDriver: testDriver,
+  );
+
+  @override
+  Future<TaskResult> run() async {
+    return inDirectory<TaskResult>(testDirectory, () async {
+      // Some initializations
+      _device = await devices.workingDevice;
+      _flutterPath = path.join(flutterDirectory.path, 'bin', 'flutter');
+
+      // Prepare the SkSL by running the driver test.
+      await _generateSkSL();
+
+      // Build the app with SkSL artifacts and run that app
+      final String observatoryUri = await _buildAndRun();
+
+      // Attach to the running app and run the final driver test to get metrics.
+      final TaskResult result = await internalRun(
+        existingApp: observatoryUri,
+      );
+
+      _runProcess.kill();
+      await _runProcess.exitCode;
+
+      return result;
+    });
+  }
+
+  Future<void> _generateSkSL() async {
+    // First, clear all old sksl.json file so they won't affect the test.
+    for (final FileSystemEntity file in Directory(testDirectory).listSync()) {
+      if (file.path.endsWith('.sksl.json')) {
+        file.deleteSync();
+      }
+    }
+
+    await super.internalRun(keepRunning: true, cacheSkSL: true);
+    final Process process = await startProcess(
+      _flutterPath,
+      <String>[
+        'attach',
+        '-d', _device.deviceId,
+      ],
+    );
+    final Stream<List<int>> broadcastOut = process.stdout.asBroadcastStream();
+    _forwardStream(broadcastOut, 'attach stdout');
+    _forwardStream(process.stderr, 'attach stderr');
+
+    final Completer<bool> attachReady = Completer<bool>();
+    final Completer<bool> skslWritten = Completer<bool>();
+
+    _transform(broadcastOut).listen((String line) {
+      if (!attachReady.isCompleted && _kObservatoryReadyRegExp.hasMatch(line)) {
+        attachReady.complete(true);
+      }
+      if (!skslWritten.isCompleted && RegExp('Wrote SkSL data to').hasMatch(line)) {
+        skslWritten.complete(true);
+      }
+    });
+
+    await attachReady.future;
+    process.stdin.write('M');
```

Comment:
I added a vm service request for this, so the stdin is unnecessary:  https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/integration.shard/vmservice_integration_test.dart#L85

Reply:
Right now the method is only hooked up in debug mode, are you running in profile/release to generate the sksl? or are you sending the request during drive?

Reply:
if its easier I could add logic to write on exit for drive

Reply:
Does this work: https://github.com/flutter/flutter/pull/58743 ?

Reply:
Yes, I'm running in profile mode. https://github.com/flutter/flutter/pull/58743 would make this PR much more easier, and it would close https://github.com/flutter/flutter/issues/54484!